main/src/ca/uwaterloo/flix/language/phase/unification/BoolMinimization.scala

```diff
@@ -0,0 +1,482 @@
+ */
+
+package ca.uwaterloo.flix.language.phase.unification
+
+import ca.uwaterloo.flix.language.ast.{Kind, Rigidity, SourceLocation, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+import scala.annotation.tailrec
+
+object BoolMinimization {
+
+  sealed trait Formula
+
+  object Formula {
+
+    case object True extends Formula
+
+    case object False extends Formula
+
+    case class Var(v: Type.KindedVar) extends Formula
+
+    case class Not(term: Formula) extends Formula
+
+    case class And(terms: List[Formula]) extends Formula
+
+    case class Or(terms: List[Formula]) extends Formula
+
+  }
+
+  /**
+    * Sort terms `True` -> `False` -> `Var`s (by name then negation) ->
+    * complex negation -> conjunctions -> disjunctions.
+    */
+  private def sortTerms(terms: List[Formula]): List[Formula] = {
+    import Formula._
+    terms.sortWith {
+      case (True, True) => false
+      case (True, False | Var(_) | Not(_) | And(_) | Or(_)) => true
+
+      case (False, True | False) => false
+      case (False, Var(_) | Not(_) | And(_) | Or(_)) => true
+
+      case (Var(_), True | False) => false
+
+      case (Var(v1), Var(v2)) => v1 < v2
+      case (Var(v1), Not(Var(v2))) if v1 == v2 => true
+      case (Var(v1), Not(Var(v2))) => v1 < v2
+      case (Not(Var(v1)), Var(v2)) if v1 == v2 => false
+      case (Not(Var(v1)), Var(v2)) => v1 < v2
+      case (Not(Var(v1)), Not(Var(v2))) => v1 < v2
+      case (Var(_), Not(_) | And(_) | Or(_)) => true
+
+      case (Not(_), True | False | Var(_)) => false
+      case (Not(_), Not(_)) => false
+      case (Not(_), And(_) | Or(_)) => true
+
+      case (And(_), True | False | Var(_) | Not(_)) => false
+      case (And(_), And(_)) => false
+      case (And(_), Or(_)) => true
+
+      case (Or(_), True | False | Var(_) | Not(_) | And(_) | Or(_)) => false
+    }
+  }
+
+  /**
+    * Reduces things like `y ∨ y` to `y` or `¬x ∧ ¬x` to `x`. Returns `None` if
+    * the list contains both `x` and `¬x` for some `x`.
+    * Assumes the list is sorted.
+    */
+  private def removeSimpleDuplicates(terms: List[Formula]): Option[List[Formula]] = {
+    import Formula._
+    terms match {
+      case Nil =>
+        Some(Nil)
+      case _ :: Nil =>
+        Some(terms)
+      case Var(v1) :: Var(v2) :: rest if v1 == v2 =>
+        removeSimpleDuplicates(Var(v2) :: rest)
+      case Not(Var(v1)) :: Not(Var(v2)) :: rest if v1 == v2 =>
+        removeSimpleDuplicates(Not(Var(v2)) :: rest)
+      case Var(v1) :: Not(Var(v2)) :: _ if v1 == v2 =>
+        None
+      case other :: rest =>
+        removeSimpleDuplicates(rest).map(other :: _)
+    }
+  }
+
+  /**
+    * Performs unit propagation on `terms` of a conjunction. If `x` or `¬y` is
+    * present, then they must be `True` and `False` respectively. This is
+    * substituted in the remaining conjunction.
+    */
+  private def unitPropagation(terms: List[Formula]): Option[List[Formula]] = {
```

Comment:
In addition to unitPropagation, we should do subsumption after.

Subsumption is when you have `(x or y) and (x or y or z)` then you can remove the last clause.

This probably requires a n^2 walk through the formula.

Reply:
The generalization of unit propagation and this is probably to carry a Formula -> True|False map substitution through the traversal?

Do you want this to done eagerly in `mkAnd` like it is now or only on the CNF transformation?

Reply:
I would keep it after cnf.

Reply:
I moved unit propagation form `mkAnd` to `toCNF` and now it has significantly slowed down